### PR TITLE
Enable thinLTO

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Enable ThinLTO for clang builds, continuation of https://github.com/ClickHouse/ClickHouse/pull/10435

Compiling/linking unitests with LTO is extremely slow and resource-intensive. On my E5-2650 dell server with 128GB ram, it takes more than 10 minutes to link, and some linking jobs even got oom killed. It's not enough to only supply `-fno-lto` at the final linking stage to disable LTO. If `lld` sees that there are bitcode files in the input it will run the lto optimizer. So we have to make the choice upon the `ENABLE_TESTS` cmake flag. 

P.S., there are multiple clang build variants in the current CI's setup. I don't know if there exists one that can be used for performance testing and with `ENABLE_TESTS=0`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -222,6 +222,20 @@ if (COMPILER_CLANG)
...
+    # We cannot afford to use LTO when compiling unitests, and it's not enough
+    # to only supply -fno-lto at the final linking stage. So we disable it
+    # completely.
+    if (NOT ENABLE_TESTS)
+        # Link time optimization
+        set (CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} -flto=thin")
+        set (CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -flto=thin")
+        set (CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO} -flto=thin")
+
+        # We cannot use GNU ar when llvm LTO is enabled
```
Comment (alexey-milovidov) [inline CMakeLists.txt:235]: Ok.

And maybe we should always prefer llvm tools when building with clang - maybe they are somewhat superior?

**Comment by alexey-milovidov** [comment]
> I don't know if there exists one that can be used for performance testing and with ENABLE_TESTS=0.

It should be the only clang/relwithdebinfo/no-sanitizer build... We'll see.

File: CMakeLists.txt (hunk 1 above)
Comment (amosbird) [reply CMakeLists.txt:235]: Yeah, that makes sense even without being superior.

**Comment by alexey-milovidov** [comment]
Or we don't have appropriate symlink to llvm-ar-10.

**Comment by alexey-milovidov** [comment]
I'll add it.

**Comment by alexey-milovidov** [comment]
Looks like it was built successfully:
https://clickhouse-builds.s3.yandex.net/10585/f11148a77c1f387158fe43f3203d47402aee557f/report.html
this is 
clang-9 | relwithdebuginfo | none | bundled | unsplitted

**Comment by alexey-milovidov** [comment]
Exception handling is broken: https://clickhouse-test-reports.s3.yandex.net/10585/f11148a77c1f387158fe43f3203d47402aee557f/functional_stateless_tests_(thread)/test_run.txt.out.log

**Comment by alexey-milovidov** [comment]
Looks like the final linking stage took 10 minutes:
```
2020-05-01 14:33:29 [7273/7283] Linking CXX executable programs/clickhouse-odbc-bridge
2020-05-01 14:33:29 ld.lld: warning: <unknown>:0:0: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering
2020-05-01 14:43:08 [7274/7283] Linking CXX executable programs/clickhouse
2020-05-01 14:43:08 ld.lld: warning: <unknown>:0:0: loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering
```

This is acceptable :)

**Comment by alexey-milovidov** [comment]
The size of
`clickhouse` was increased from 258 858 K to 276 878 K
`clickhouse-odbc-bridge` was lowered from 94 481 K to 86 246 K.

**Comment by alexey-milovidov** [comment]
At least it works.

**Comment by alexey-milovidov** [comment]
On basic set of queries from https://clickhouse.tech/benchmark/hardware/ there is no visible difference.

**Comment by amosbird** [comment]
I guess vectorized processing won't benefit much from cross TU optimizations. Some micro-optimizations are required to close the performance gap.

**Comment by alexey-milovidov** [comment]
Looks like there is some difference but my setup is noisy...
At least it is still slower than gcc.

**Comment by alexey-milovidov** [comment]
> Some micro-optimizations are required to close the performance gap.

There are many cases when we put functions in headers for inlining.
LTO can eliminate the need for it and find more cases where inlining can improve code.

**Comment by amosbird** [comment]
> There are many cases when we put functions to headers for inlining.

Yeah, but I think we've already covered most inlinable cases that can produce performance benefits.

One thing I noticed can have noticeable effects on performance when doing micro-optimizations is the register allocation algorithm. And another famous fact is the instruction layout https://www.youtube.com/watch?v=r-TLSBdHe1A  :)

**Comment by alexey-milovidov** [comment]
[benchmark_gcc_clang.html.txt](https://github.com/ClickHouse/ClickHouse/files/4566306/benchmark_gcc_clang.html.txt)

Screenshot_20200502_001423

ThinLTO can cut half of the gap between gcc and clang.
We need to also try with clang-10.

**Comment by alexey-milovidov** [comment]
Also remember that we don't omit-frame-pointer in clang builds.

**Comment by amosbird** [comment]
> Also remember that we don't omit-frame-pointer in clang builds.

Let's try if it can close the gap now :)

**Comment by alexey-milovidov** [comment]
Ok. And this PR is ready to merge.

**Pull request merged**